mono/metadata/icalls.h

```diff
@@ -0,0 +1,805 @@
+#ifndef __MONO_METADATA_ICALLS_H__
+#define __MONO_METADATA_ICALLS_H__
+
+/* From MonoProperty.cs */
+typedef enum {
+	PInfo_Attributes = 1,
+	PInfo_GetMethod  = 1 << 1,
+	PInfo_SetMethod  = 1 << 2,
+	PInfo_ReflectedType = 1 << 3,
+	PInfo_DeclaringType = 1 << 4,
+	PInfo_Name = 1 << 5
+} PInfo;
+
+#include "appdomain-icalls.h"
+#include "console-io.h"
+#include "environment.h"
+#include "environment-internal.h"
+#include "file-mmap.h"
+#include "filewatcher.h"
+#include "gc-internals.h"
+#include "locales.h"
+#include "monitor.h"
+#include "mono-perfcounters.h"
+#include "mono-route.h"
+#include "object-internals.h"
+#include "rand.h"
+#include "security.h"
+#include "security-core-clr.h"
+#include "security-manager.h"
+#include "string-icalls.h"
+#include "sysmath.h"
+#include "threadpool.h"
+#include "threadpool-io.h"
+#include "threads-types.h"
+#include "utils/mono-digest.h"
+#include "utils/mono-proclib.h"
+#include "utils/mono-time.h"
+#include "w32event.h"
+#include "w32file.h"
+#include "w32mutex.h"
+#include "w32process.h"
+#include "w32semaphore.h"
+#include "w32socket.h"
+
+ICALL_EXPORT MonoReflectionAssemblyHandle
```

Comment:
So there's now a declaration for every icall, right?

If so, why can't we just use the standard `#ifdef __cplusplus`/`export "C" {` mechanism (in its `G_BEGIN_DECLS`) for this whole header instead of mucking about trying to get `ICALL_EXPORT` past GCC's parser.

Reply:
Yes we could do that. You lose typesafe linkage, but on these functions that is probably ok, or even required (the taking address in the table would be ambiguous and fail to compile).

Reply:
(You'd have to cast to disambiguate, which is kinda horrible at first, not sure there is a good way.)

Reply:
Then ICALL_EXPORT needs to be extern "C", and have to fix all the missing ones.
Will try.

Reply:
I kinda still would prefer to minimize extern "C". It seems more "pure" to avoid it. But ok it is mostly done locally.